zebrad/src/components/mempool/crawler/tests.rs

```diff
@@ -72,7 +67,7 @@ proptest! {
                 recent_sync_lengths.push_extend_tips_length(sync_length);
             }
 
-            Ok(())
+            Ok::<(), TestCaseError>(())
         })?;
```

Comment:
Ergononics Nitpick: It's unfortunate that the `Result` type isn't inferred.

Is this a limitation of Rust's current async type inference implementation?
Does the compiler give a helpful error if the turbofish (`::<(), TestCaseError>`) is removed?
Does removing the outer try operator (`?`) make type inference work?

This isn't a blocker, it's just something to consider when evaluating the design.

Reply:
Yeah, I think the problem is the type inference for the `async` block. From what I understood, it should infer the `Future::Output` type to be `Result<(), E>`, where `E: From<TestCaseError>`. Even though the [closure generated by the `proptest!`](https://github.com/AltSysrq/proptest/blob/master/proptest/src/sugar.rs#L955-L960) macro [expects a `TestCaseError`](https://docs.rs/proptest/1.0.0/proptest/test_runner/struct.TestRunner.html#method.run), Rust still can't infer it completely because the outer `?` just means that the constraint `TestCaseError: From<E>` must be true.

The error it gives isn't that bad, but it's not obvious what needs to be done, because it suggests adding an explicit type for the `async` block `Future` instead of just the error:

```
error[E0282]: type annotations needed
  --> zebrad/src/components/mempool/crawler/tests.rs:70:13
   |
70 |             Ok(())
   |             ^^ cannot infer type for type parameter `E` declared on the enum `Result`
   |
help: consider specifying the type argument in the method call
   |
34 |         runtime.block_on::<F>(async move {
   |                         ^^^^^
```

Removing the outer `?` doesn't change anything.